FIX: restore b2g agent sheet scrollbars and other mobile tweaks on b2g windows

rpl opened the pull request:
- new DOM window detected by notification observers
- inject full original b2g agent sheet css
  (e.g. to style buttons as in the FirefoxOS device)

should fix #379 

NOTE: there're some remaining issues related to shell styling, 
because currently any try to inject b2g agent sheet into the homescreen
iframe generates some weird side-effect (homescreen change rotation
instead to lock to portrait) but probably it's more important to restore the style for apps now, and then in a second time try to restore it on the homescreen as well.

rpl committed:
reinject b2g mobile agent stylesheet from a "content-document-global-created" notification observer

rpl commented:
Screenshots of a couple of apps with reinjected b2g stylesheet:

![app_with_mobile_scrollbar](https://f.cloud.github.com/assets/11484/320548/1feb28c2-994b-11e2-83ff-28583c70c79a.png)
![firefoxos_browser_mobile_stylesheet](https://f.cloud.github.com/assets/11484/320549/203336d0-994b-11e2-83fd-8260bfc0d130.png)

rpl commented:
Screenshots of a remaining issues on homescreen stylesheet:

![style_issue_on_b2g_shell_style1](https://f.cloud.github.com/assets/11484/320553/4de729d8-994b-11e2-9ec9-685a24229c32.png)
![style_issue_on_b2g_shell_style2](https://f.cloud.github.com/assets/11484/320554/4dee4d1c-994b-11e2-85ab-775473620379.png)

mykmelez commented:
_FloatingScrollbars.jsm_ looks like it is designed to enable a stylesheet to be added and removed on demand during the lifetime of a content frame. In this case, however, we only need to add a stylesheet once, when a content frame is first created; and we don't need to remove it.

So we should be able to do something simpler, like observing content window creation and adding a stylesheet to each content window that gets created. Here's a self-contained example that works in my tests, including for the homescreen:

```
Services.obs.addObserver(
  function injectStylesheet(subject, topic, data) {
    debug("injectStylesheet: " + data);
    let URL = Services.io.newURI("chrome://prosthesis/skin/b2g-content.css",
                                 null, null);
    let winUtils = subject.QueryInterface(Ci.nsIInterfaceRequestor).
                   getInterface(Ci.nsIDOMWindowUtils);
    winUtils.loadSheet(URL, winUtils.AGENT_SHEET);
  },
  "content-document-global-created",
  false
);
```

It should be straightforward to integrate that into _SimulatorWindowManager_, so we shouldn't need a separate _B2GAgentSheet.jsm_ module.

Also, I think _content-document-global-created_ will suit our purpose just fine, but note the others listed on [Observer Notifications](https://developer.mozilla.org/en-US/docs/Observer_Notifications). If we run into problems adding the stylesheet to the right set of windows at the right time, then there are other options.

mykmelez commented on prosthesis/skin/b2g-content.css:
```diff
@@ -0,0 +1,289 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Instead of putting this into the _skin/_ directory and calling it _b2g-content.css_, put it into the _content/_ directory and call it simply _b2g.css_. So its URL will be _chrome://prosthesis/content/b2g.css_. Over in issue #343 I plan to remove the _skin/_ directory; in the meantime, we shouldn't add anything to it anymore.

rpl committed:
inject into the FirefoxOS iframe container to fix style issues on app switcher and mobile select

rpl commented:
> So we should be able to do something simpler, like observing content window creation and adding a stylesheet to each content window that gets created.

@mykmelez awesome! it's much more simple and clean now.

I've just rebased this pull request on the current master and updated this pullrequest.

I wasn't aware of this notification, in my tests it's working as well as the previous prototype (using less code :-P),
it doesn't seem to be able to solve the remain corner cases. 

In 4e18ea9 I've tweaked the suggested fix, injecting b2g.css stylesheet into the main FirefoxOS iframe 
to fix the remain style issues (e.g. app switcher and mobile select)

mykmelez commented on prosthesis/content/window-manager.js:
```diff
@@ -52,6 +55,24 @@ let SimulatorWindowManager = {
+  },
+  _injectStylesheet: function(win) {
+    let URL = Services.io.newURI("chrome://prosthesis/content/b2g.css",
+                                 null, null);
```
Nit: as a variable, this name should be lower-case, i.e. _url_ (even though I mistakenly used upper-case in the original example!). Alternately, you could make this a global const like _B2G_STYLESHEET_URL_, since it never changes.

mykmelez commented on prosthesis/content/window-manager.js:
```diff
@@ -52,6 +55,24 @@ let SimulatorWindowManager = {
+    let URL = Services.io.newURI("chrome://prosthesis/content/b2g.css",
+                                 null, null);
+    let winUtils = win.QueryInterface(Ci.nsIInterfaceRequestor).
+      getInterface(Ci.nsIDOMWindowUtils);
```
Nit: indent this as:

```
    let winUtils = win.QueryInterface(Ci.nsIInterfaceRequestor).
                   getInterface(Ci.nsIDOMWindowUtils);
```

mykmelez commented:
> I wasn't aware of this notification, in my tests it's working as well as the previous prototype (using less code :-P),
> it doesn't seem to be able to solve the remain corner cases. 

Which corner cases remain? It looks like you've fixed the homescreen issues. Or are there still some issues with the homescreen?

rpl committed:
minor tweaks on prothesis window-manager.js

rpl committed:
prevents b2g stylsheet inject side-effect on running an app during b2g startup

- move b2g stylesheet injection code into its own file (inject-b2g-agentsheet.js)
- observe "content-document-global-created" notification before ContentStart to
  inject into "iframe#homescreen" created inside shell.xul
  (prevents layouting bug on running app during b2g startup)

rpl commented:
> Which corner cases remain? It looks like you've fixed the homescreen issues. Or are there still some issues with the homescreen?

@mykmelez my previous try (to fix the stylesheet issues in app switcher or mobile select) introduced a random 
rendering/layouting bug when we run an app on a starting b2g instance, e.g. pressing the run button when
b2g desktop is not already running)

I fixed the bug and removed the side-effect in the last commit (91d85de):
- move b2g stylesheet injection code into its own file (inject-b2g-agentsheet.js)
- observe "content-document-global-created" notification before ContentStart to
  inject into "iframe#homescreen" created inside shell.xul
  (prevents layouting bug on running app during b2g startup)

now it should work correctly in all the cases (at least in the known cases)

mykmelez commented on prosthesis/content/inject-b2g-agentsheet.js:
```diff
@@ -0,0 +1,15 @@
+const B2G_AGENTSHEET_URL = Services.io.newURI("chrome://prosthesis/content/b2g.css",
```
This line exceeds 80 characters, so it would be better to format the statement as:

```
const B2G_AGENTSHEET_URL =
  Services.io.newURI("chrome://prosthesis/content/b2g.css", null, null);
```

Also, this file needs a license header.

mykmelez commented:
Looks good! (modulo the minor issues in the previous comment)

Ideally, we would not create a separate JavaScript file for this small amount of code, but there isn't currently another file that could incorporate it sensibly. It would make sense to put it into shell.js, but we load that on ContentStart at the moment, because months ago we had trouble loading it early. But we should try that again, since the bug I encountered, which I didn't have time to investigate further back then, might well have been fixed by now.

mykmelez commented:
I'm going to fix the last two issues and then merge this so I can get it into the preview builds I'm about to spin.
